lymph/unilateral.py

```diff
@@ -708,21 +662,142 @@ def modalities(self, modality_spsn: Dict[Any, List[float]]):
             1 - s_P & s_N
             \\end{pmatrix}
         """
-
         if hasattr(self, "_observation_matrix"):
             del self._observation_matrix
         if hasattr(self, "_obs_list"):
             del self._obs_list
+        if not hasattr(self, "_spsn_tables"):
+            self._spsn_tables = {}
+
+        dictionary = modality_spsn
+        for mod,matrix in dictionary.items():
+            self._spsn_tables[mod] = matrix
+
+
+    def _check_modality(self, modality: str, spsn: list):
+        """Private method that checks whether all inserted values
+        are valid for a confusion matrix.
+
+        Args:
+            modality (str): name of the modality
+            spsn (list): list with specificity and sensiticity
+
+        Raises:
+            TypeError: returns a type error if the modality is not a string
+            ValueError: raises a value error if the spec or sens is not a number btw. 0.5 and 1.0
+        """
+        if not isinstance(modality, str):
+                msg = ("Modality names must be strings.")
+                raise TypeError(msg)
+        has_len_2 = len(spsn) == 2
+        is_above_lb = np.all(np.greater_equal(spsn, 0.5))
+        is_below_ub = np.all(np.less_equal(spsn, 1.))
+        if not has_len_2 or not is_above_lb or not is_below_ub:
+            msg = ("For each modality provide a list of two decimals "
+                "between 0.5 and 1.0 as specificity & sensitivity "
+                "respectively.")
+            raise ValueError(msg)
+
+    def clinical(self, modality_spsn: Dict[Any, List[float]]) -> Dict:
+        """produces the confusion matrix of a clinical modality, i.e. a modality
+        that can not detect microscopic metastases
+
+        Args:
+            modality_spsn (Dict[Any, List[float]]): dictionary with modality name and [sp,sn] list
+
+        Returns:
+            Dict[str,np.ndarray]: returns a dictionary with modality name and clinical confusion matrices
+        """
+        modality_dictionary = {}
+        for mod, spsn in modality_spsn.items():
+            try:
+                self._check_modality(mod,spsn)
+            except TypeError:
+                raise
+            except ValueError:
+                raise
+
+            sp, sn = spsn
+            modality_dictionary[mod] = np.array([[sp, 1. - sp],
+                                               [sp, 1. - sp],
+                                               [1. - sn, sn]])
+        return modality_dictionary

```

Comment:
This is nice! But I think it would be even more useful if it were not a method of the `Unilateral` class, but a normal function outside of it.

Then, the function signature could look like this:
```python
def clinical(specificity: float, sensitivity: float) -> np.ndarray:
    """Create a 2x3 diagnostic matrix for a clinical modality."""
    # Some checks ...
    return np.array([...])
```
And then the modalities setter inside the `Unilateral` class always expects a 2x3 matrix for any modality (or 2x2, in case of binary). That way it's completely independent.

Reply:
This can be done, but the `confusion_matrix` option would need to stay inside `Unilateral` if we want to check whether it conforms to the number of allowed states (here we need access to `self.allowed_lnl_states`).  Alternatively, the check could be done inside the setter method, which would be no problem except that a potential wrong input is only caught farther down the code.

Reply:
You can just check the dimensions of the matrix that is passed to the setter in the modalities dictionary. Which is probably necessary anyways, since we want to allow custom modalities, no?

Reply:
yes would work. I did all the checking inside the clinical/pathological function so far. Such that any "error" is caught as early as possible. But it doesn't even matter, since the modalities are only set once and not during the sampling --> it's not computationally expensive.